p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
+// Accepted() to allow the router to set up its internal queues before reactors
+// start sending messages (holding the Router.peerMtx mutex while calling
+// Accepted or Dialed will halt all message routing while peers are set up, which
+// is too expensive and also causes difficulties in tests where we may want to
+// consume peer updates and send messages sequentially).
+//
+// FIXME: This possibly indicates an architectural problem. Should the peerManager
+// handle actual network connections to/from peers as well? Or should all of this
+// be done by the router?
+func (m *peerManager) Ready(peerID NodeID) {
+	m.mtx.Lock()
+	connected := m.connected[peerID]
+	m.mtx.Unlock()
+	if connected {
+		m.broadcast(PeerUpdate{
+			PeerID: peerID,
+			Status: PeerStatusUp,
+		})
+	}
+}
+
+// Disconnected unmarks a peer as connected, allowing new connections to be
+// established.
+func (m *peerManager) Disconnected(peerID NodeID) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	delete(m.connected, peerID)
+	m.broadcast(PeerUpdate{
+		PeerID: peerID,
+		Status: PeerStatusDown,
+	})
+	return nil
+}
+
+// peerStore stores information about peers. It is currently a bare-bones
+// in-memory store, and will be fleshed out later.
+//
+// peerStore is not thread-safe, since it assumes it is only used by peerManager
+// which handles concurrency control. This allows multiple operations to be
+// executed atomically, since the peerManager will hold a mutex while executing.
+type peerStore struct {
+	peers map[NodeID]peerInfo
+}
+
+// newPeerStore creates a new peer store.
+func newPeerStore() *peerStore {
+	return &peerStore{
+		peers: map[NodeID]peerInfo{},
 	}
+}
+
+// Get fetches a peer, returning nil if not found.
+func (s *peerStore) Get(id NodeID) (*peerInfo, error) {
 	peer, ok := s.peers[id]
 	if !ok {
-		peer = newStorePeer(id)
-		s.peers[id] = peer
+		return nil, nil
 	}
-	s.claimed[id] = true
-	return peer
+	return &peer, nil
 }
 
-// Dispense finds an appropriate peer to contact and claims it. The caller has
-// exclusive ownership of the peer, and must return it by calling Return(). The
-// peer will not be dispensed again until returned.
-//
-// Returns nil if no appropriate peers are available.
-func (s *peerStore) Dispense() *peerInfo {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	for key, peer := range s.peers {
-		switch {
-		case len(peer.Addresses) == 0:
-		case s.claimed[key]:
-		default:
-			s.claimed[key] = true
-			return peer
-		}
+// Set stores peer data.
+func (s *peerStore) Set(peer *peerInfo) error {
+	if peer == nil {
+		return errors.New("peer cannot be nil")
 	}
+	s.peers[peer.ID] = *peer
 	return nil
 }
 
-// Return returns a claimed peer, making it available for other
-// callers to claim.
-func (s *peerStore) Return(id NodeID) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	delete(s.claimed, id)
+// List retrieves all peers.
+func (s *peerStore) List() ([]*peerInfo, error) {
```

Comment:
Is there a reason to return references to `peerInfo` objects?

Reply:
Just to avoid copying the struct all over the place. This needs to be revisited, since we have to make sure the caller can't inadvertently change data in an in-memory database or in other callers, but we'll need to explicitly copy in that case anyway since e.g. slices and maps aren't copied automatically.